Make the Preview Menu extensible

delawski opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

This PR introduces a new `<PluginPreview>` component to the `@wordpress/block-editor` package. It allows for defining a custom "Preview" menu item along with optional content that is rendered instead of the `VisualEditor`.

Thanks to that, plugin authors can extend the "Preview" menu and provide optional custom content previews.

This PR is based on #25430 proposed by @mreishus. It addresses @jorgefilipecosta's [feedback](https://github.com/WordPress/gutenberg/pull/25430#discussion_r503445235) pointing out that the new component added to `@wordpress/block-editor` should not depend on the `@wordpress/edit-post` package.

In the original implementation we were using a `deviceType` and a `setDeviceType` action from the `core/edit-post` store. Right now, they are [passed](https://github.com/WordPress/gutenberg/compare/trunk...delawski:add/preview-menu-slot#diff-61a1980577fb29dd7a7ed4d0a4c23919d5c6bbae93517a9957f80e1c04d50392R106-R107) to the slot fill by the [`VisualEditorOrPluginPreview`](https://github.com/WordPress/gutenberg/compare/trunk...delawski:add/preview-menu-slot#diff-4ea0366d68b391dbc2c4825aeba64fc12057883203c03bfe0a50c9ebecf4e51bR28) component (that lives in the `@wordpress/edit-post` package).

There's one additional difference between this PR and the [original one](#25430). The custom preview (`PluginPreview` component children) is optional now. If there is no custom preview provided, selecting a custom "Preview" menu item will not affect the `VisualEditor`. Instead, plugin authors may rely on the `onClick` event handler for providing custom functionality. This may be very helpful in the case of plugins that provide external previews (e.g. paywalled content, search results, or AMP preview as proposed in ampproject/amp-wp#5943).

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
- Create a new block editor plugin as follows:

```js
import domReady from '@wordpress/dom-ready';
import { __ } from '@wordpress/i18n';
import { PluginPreview } from '@wordpress/block-editor';
import { registerPlugin } from '@wordpress/plugins';
import { external } from '@wordpress/icons';

domReady( () => {
	const PluginPreviewTest = () => (
		<>
			<PluginPreview
				name="preview-custom-1"
				title={ __( 'Custom Preview 1' ) }
			>
				<h1>
					{ __( 'Custom Preview 1 Content' ) }
				</h1>
			</PluginPreview>

			<PluginPreview
				name="preview-custom-2"
				title={ __( 'Custom Preview 2' ) }
				onClick={ ( event ) => console.log( event  ) }
			>
				<h1>
					{ __( 'Custom Preview 2 Content' ) }
				</h1>
			</PluginPreview>

			<PluginPreview
				name="custom-action"
				title={ __( 'Custom Action' ) }
				icon={ external }
				onClick={ ( event ) => console.log( event ) }
			/>
		</>
	);

	registerPlugin( "plugin-preview-test", {
		render: PluginPreviewTest,
	} );
} );
```
- Go to block editor and confirm there are 3 new items in the "Preview" menu:
![Screenshot 2021-05-19 at 11 38 23](https://user-images.githubusercontent.com/478735/118791168-bff45c80-b896-11eb-80f5-55e7e73ccc90.png)
- Click the "Custom Preview 1" menu item and confirm that the block editor content has been replaced with a custom preview
- Click the "Custom Preview 2" menu item and confirm that besides changing the editor content, a click event object has been logged to the console
![Screenshot 2021-05-19 at 11 40 19](https://user-images.githubusercontent.com/478735/118791450-034ecb00-b897-11eb-93bf-f351071df650.png)
- Click the "Custom Action" menu item and confirm the editor config has not been changed while a click event has been logged to the console.

## Screenshots <!-- if applicable -->

![Screenshot 2021-05-18 at 20 50 27](https://user-images.githubusercontent.com/478735/118790681-4bb9b900-b896-11eb-8828-035af3a5f0c2.png)

## Types of changes
New feature (non-breaking change which adds functionality)

Fixes #25309
Relates to Automattic/wp-calypso#42896
Relates to ampproject/amp-wp#5943

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

mreishus committed:
Allow Preview menu to be extendible by plugins: WIP

mreishus committed:
Allow Preview menu to be extendible by plugins: WIP

mreishus committed:
PluginPreview{MenuItem,}: Change external interface "deviceName" -> "previewId"

mreishus committed:
PluginPreview{MenuItem,}: Change external interface "deviceName" -> "previewId"

delawski committed:
Decouple PluginPreviewMenuItem from @wordpress/edit-post

delawski committed:
Decouple PluginPreviewMenuItem from @wordpress/edit-post

delawski committed:
Retrieve previewId in parent component

delawski committed:
Retrieve previewId in parent component

delawski committed:
Accept custom onClick handler and icon in plugin preview

delawski committed:
Accept custom onClick handler and icon in plugin preview

delawski committed:
Update jsdoc blocks

delawski committed:
Update jsdoc blocks

delawski committed:
Check previewId aka deviceType directly in child component

delawski committed:
Check previewId aka deviceType directly in child component

delawski committed:
Update coreDeviceTypes docs

delawski committed:
Update coreDeviceTypes docs

delawski committed:
Update slotfills docs

delawski committed:
Update slotfills docs

delawski commented:
cc @simison @westonruter @pierlon

jasmussen commented:
Thanks for working on this! It's non-trivial to test, I got this far:

![after](https://user-images.githubusercontent.com/1204802/119095210-484b3c80-ba12-11eb-97fa-2efae954bb06.gif)

... but have to move on to some other tasks, so can't explore the intricacies too deeply. For that reason, thank you for including great screenshots as well.

High level, I still think it makes a lot of sense for us to allow this menu to be extensible, so that plugins can add custom preview interfaces exactly like what is outlined for AMP. 

Looking at the current API, though, it seems like this would plugins to add anything there — not just menu items, which I fear is likely to be abused to show all sorts of irrelevant things. Is there any way we can rein in the API to only allow menu items in a very specific format? Ideally plugin developers should not have to know about which components to use to make a menu item look correctly like a menu item. Considering we can always expand this in the future, what is the least we can start with? Label & onclick action? Label & external link action? Do we need to allow custom icons & subgroups?

The less we can start with, the better we'll know where to improve and expand upon it in future iterations. What do you think?

gziolo commented:
@jasmussen, should we follow the same path as with the `PluginSidebar` with the automatic menu item injection? We could even go one step further and don't expose the menu item fill at all, so all you would have to do is to define `PluginPreview` and it would know where to render the menu item?

By the way, isn't the name "preview" too generic?

jasmussen commented:
> We could even go one step further and don't expose the menu item fill at all, so all you would have to do is to define PluginPreview and it would know where to render the menu item?

I like the sound of that. While as noted I think there's a place for this feature, I like for APIs to be mostly agnostic in their implementation, so if we were to theoretically move the preview dropdown somewhere else in the interface, the API would still work and make sense. 

> By the way, isn't the name "preview" too generic?

Good point, especially as I've seen recent conversations about renaming it to just "View".

delawski committed:
Do not expose menu item fill; simplify PluginPreview component

As per the review feedback, there seems to be no need to expose the Preview menu item fill at all. It should be sufficient to expose the title, icon, and click handler. This way we "protect" the Preview menu drop-down from injecting arbitrary, irrelevant elements into it.

Thanks to that, the entire custom previews implementation could be simplified. It's now mostly contained within the `PluginPreview` component itself.

delawski committed:
Do not expose menu item fill; simplify PluginPreview component

As per the review feedback, there seems to be no need to expose the Preview menu item fill at all. It should be sufficient to expose the title, icon, and click handler. This way we "protect" the Preview menu drop-down from injecting arbitrary, irrelevant elements into it.

Thanks to that, the entire custom previews implementation could be simplified. It's now mostly contained within the `PluginPreview` component itself.

delawski commented:
@gziolo @jasmussen Thanks for your feedback!

> > We could even go one step further and don't expose the menu item fill at all, so all you would have to do is to define PluginPreview and it would know where to render the menu item?

> I like the sound of that. While as noted I think there's a place for this feature, I like for APIs to be mostly agnostic in their implementation, so if we were to theoretically move the preview dropdown somewhere else in the interface, the API would still work and make sense.

I like this approach, too. I think having a similar API as the `PluginSidebar` makes much sense. Can you please review the revisited implementation (after f949647)?

> > By the way, isn't the name "preview" too generic?

> Good point, especially as I've seen recent conversations about renaming it to just "View".

I haven't done anything about this, yet. Do you think `preview` won't last and we should go with a more generic `view` instead? Or should we leave it for now?

delawski commented:
Note that the PR description has been updated to reflect the changes introduced in f949647 (i.e. that there is now just one new component, the `PluginPreview`).

gziolo commented:
@youknowriad, what would be the best place to put this new SlotFill for plugins, `@wordpress/interface`?

delawski commented:
@gziolo @jasmussen Do you have any update on this one? Is there some way I could facilitate pushing it forward?

jasmussen commented:
Thanks for the ping. I know that many of the folks pinged on this thread have been fever busy with WordPress 5.8 beta release things. I'm sure they'll get back to you as soon as they get a breather. Thank you for your patience and your work!

mtias commented:
Is this blocked? Let's pick it up :)

gziolo commented on packages/block-editor/src/components/index.js:
```diff
@@ -89,7 +89,11 @@ export { default as withColorContext } from './color-palette/with-color-context'
 export { default as __unstableBlockSettingsMenuFirstItem } from './block-settings-menu/block-settings-menu-first-item';
 export { default as __unstableInserterMenuExtension } from './inserter-menu-extension';
-export { default as __experimentalPreviewOptions } from './preview-options';
+export {
```
Can you explain why `__experimentalPreviewOptions` and `coreDeviceTypes` are proposed to become part of public API? They don't seem to be used outside of the package where they are defined.

gziolo commented on packages/block-editor/src/components/plugin-preview/index.js:
```diff
@@ -0,0 +1,78 @@
+ *                                   that provide no content (`children`).
+ * @param {string}    props.title    Menu item title.
+ */
+export default function PluginPreview( {
```
As noted in my previous comment, I don't think that this SlotFill should be located inside `@wordpress/block-editor` as this package contains APIs only related to the block editing interface. This is a higher-level concept that in fact replaces the regular interface you see for editing blocks. That's why I proposed that this functionality goes to `@wordpress/interface` that already holds similar parts of the header UI like `PinnedItems` or everything related to `ComplementaryArea` that is a internal component used by `PluginSidebar`.

gziolo commented on packages/block-editor/src/components/plugin-preview/index.js:
```diff
@@ -0,0 +1,78 @@
+	const previewSlotName = `core/block-editor/plugin-preview/${ name }`;
+	const previewSlot = useSlot( previewSlotName );
+
+	if ( coreDeviceTypes.includes( name ) ) {
```
Does it mean that custom plugins can't use the existing names like `Tablet`?

gziolo commented on packages/block-editor/src/components/plugin-preview/index.js:
```diff
@@ -0,0 +1,78 @@
+					<MenuItem
+						onClick={ ( ...args ) => {
+							if ( name && previewSlot.fills?.length > 0 ) {
+								setDeviceType( name );
```
In the examples provided with the documentation, I see for example `preview-custom-1`. Is it going to be set as the device type? If that's true then `setDeviceType` doesn't seem to be a good fit anymore here and it should be replaced with something more general.

gziolo commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -225,7 +225,7 @@ function Layout( { styles } ) {
 						) }
 						{ isRichEditingEnabled && mode === 'visual' && (
-							<VisualEditor styles={ styles } />
+							<VisualEditorOrPluginPreview styles={ styles } />
```
It feels like `PluginPreview` should be independent of the `mode`. Well, maybe at least it should still be possible to see the preview when `isRichEditingEnabled` is disabled. It needs some more thinking because depending on the content of the preview, it might not play with how we use the visual editor today.

Could you disable rich editing, use text mode to edit content and still use the custom preview from the plugin?

gziolo reviewed (commented):
I left some feedback to discuss. It's a great start as we have a working prototype. We need to clarify now how it's going to work and how to optimize the usage of public APIs.

Buddierdl commented:
Hi @delawski, has there been any progress on this? I would love to have this feature merged.

delawski commented:
Hey @Buddierdl, unfortunately not. I've been busy with some other tasks but I'm planning to get back to it near term.

simison commented:
@delawski hi! Any updates on this?

arcangelini committed:
Merge branch 'add/preview-menu-slot' of https://github.com/delawski/gutenberg into add/preview-menu-slot

arcangelini committed:
Merge remote-tracking branch 'upstream/trunk' into add/preview-menu-slot

leutrimhusaj commented:
Is it planned to add more devices?
Maybe have 5 devices 'Desktop', 'Tablet Landscape', 'Tablet horizontal', 'Mobile Landscape', 'Mobile horizontal'.
So the new devices appear inside the first MenuGroup as a MenuItem.

gziolo commented:
Noting here that we are discussing a general strategy for making the editor interface extensible in #37448.

gziolo commented:
Implemented with #64644. Thank you for your contribution.
